New activity budget and FFP section

mgwalker opened the pull request:
### This pull request changes...

- switches to the new costs table and FFP design
- changes the print view to be one great big costs table, with other funding and FFP included
- updates internal state data tools
- adds tests
- closes #2074

### This pull request is ready to merge when...

- [x] Tests have been updated (and all tests are passing)
- [x] This code has been reviewed by someone other than the original author
- [ ] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented
- N/A ~The change has been documented~
- [x] Changelog is updated as appropriate

### This feature is done when...

- [ ] Design has approved the experience
- [ ] Product has approved the experience

mgwalker commented:
@hbillings Would it make sense for you to do a code review on this before taking it on to get it across the finish line?

mgwalker commented on web/src/containers/activity/CostAllocateFFP.js:
```diff
@@ -1,276 +1,369 @@
+    getActivity = selectActivityByIndex,
+    getCostAllocation = selectCostAllocationForActivityByIndex,
+    getCostSummary = selectActivityCostSummary,
+    getState = getUserStateOrTerritory
```
Made these injectable so the tests would be easier.

mgwalker commented on web/src/reducers/activities.selectors.js:
```diff
@@ -87,10 +187,31 @@ export const selectActivitySchedule = createSelector(
+            return -1;
+          }
+
+          return 0;
```
Here's that schedule sorting fix

hbillings replied:
nice!

hbillings commented on web/src/reducers/activities.selectors.js:
```diff
@@ -30,31 +30,131 @@ export const selectAllActivities = ({
+    { activityIndex }
+  ) =>
+    years.reduce(
+      (o, ffy) => ({
```
What is `o`? Could we maybe add a comment?

mgwalker replied:
It's the accumulator object for the reducer. I never know what to call it... It doesn't have an inherent name, it's just the temporary thing being built up. But I guess in this case, it's the key personnel, so that seems like a good and decent name for it.

mgwalker commented:
Listening to it from a screen reader, this is pretty bad. I've got an idea for how to make it a better experience by hiding the table from screen readers and instead having each row of the table become a single element that gets read all at once. Will try that out tomorrow!

jeromeleecms commented:
Super excited about this one! Just reviewing it, it's a massive amount of work 🥇 

Some things I noticed:

1. The table label doesn't updated based on the allocated amount (the math updates based on the selection in the drop down): 
![image](https://user-images.githubusercontent.com/34279968/76802338-b5bb6780-6794-11ea-9c33-d99c616f8a68.png)

2. The summary also doesn't react to a change in the dropdown (currently selected at 75/25): 

![image](https://user-images.githubusercontent.com/34279968/76802551-26fb1a80-6795-11ea-95b4-0e326ec9ad26.png)

On a related note, I totally missed your question in the original issue, for a split Federal / Share by FFY, I would suggest:

"The total cost of the Program Administration activity is $5,459,076. **Because of $338,410 in other funding**, the total cost to Medicaid is $5,120,666. This activity is using a **90/10 funding split (FFY 2019 and FFY 2020) and a 75/25 funding split (FFY 2021)** resulting in a federal share of $4,608,600 and a Missouri share of $512,066."

3. Are the column headers removed on purpose (comparing this to the mockup)? 

4. Does the alignment seem off to you with the Total column?

![image](https://user-images.githubusercontent.com/34279968/76802884-dc2dd280-6795-11ea-9390-f026abc5ef85.png)

5. I know the mockup of the original table doesn't have it, but can/should there be a dollar value included for the % time employees in the table? In the example from Activity 1, it seems like the person is 30% time but $0 which is a confusing.

![image](https://user-images.githubusercontent.com/34279968/76803037-2ca53000-6796-11ea-8bb4-f85094423b21.png)

6. Maybe a super minor detail, but I noticed the indent between the various new tables seems a little off: 

![image](https://user-images.githubusercontent.com/34279968/76803415-0633c480-6797-11ea-9af4-980df901d925.png)

jeromeleecms commented:
Separately -- this is looking GREAT. Very easy to follow the math.

mgwalker commented:
1 & 2: Oooops. I felt like I was forgetting to hook something up but I couldn't remember what it was. Thanks for the eagle eyes!

3: Yes, per [Laura's comment](https://github.com/18F/cms-hitech-apd/issues/1792#issuecomment-587949024) on the original design issue. Her comments didn't make it into the mockup, but they're capture in the "this task is done" checklist.

4. It's off, but it's behaving according to our standard rules for tables. Header cells are left-aligned, numeric cells are right-aligned. That column is pretty wide, it turns out, so the header is way to the left of the content. I will play around with getting the columns to shrink down to fit their contents. That should also help tighten up the whole table.

5. The APD key personnel are funky because we ask for their cost attributable to the program, not their total cost to the state, and we separately ask for their time commitment. The two numbers are not mathematically correlated. In the data used for the preview site, the second key person has an attributable cost of $0 but a 30% time commitment, so that's what the table shows. We don't have anything that would be the equivalent of a "unit cost" like we do with the activity state personnel, so I left that column blank. Maybe in the future we should change what we ask for APD key personnel to be in line with what we ask for activity state personnel?

6. It's because they're separate tables and I haven't tried to set any of the column widths. As a result, they grow or shrink however the browser thinks is most appropriate based on the content. I can give that first blank column a fixed width, though, and that'll at least align the names of things. In viewonly/print mode, this isn't an issue because it's rendered as one big table.

jeromeleecms commented:
Sounds good!

For your question on (5), isn't this equivalent to Unit Costs?

![image](https://user-images.githubusercontent.com/34279968/76884759-6b40f600-683b-11ea-8a7a-031adc69e12c.png)

In the case the user selects "No" - can we insert "$0" as the unit cost?

mgwalker commented:
In the math, we have always treated the key personnel cost as the total cost attributable to the program, not the key person's total cost to the *state*. (The yes/no radio buttons kind of imply that's what we're asking for, too.) So if someone enters `$50,000` and `50%` commitment, we use $50,000 for their cost in all the math. We don't scale it based on their percent commitment.

So one of two things seems to be true:

1. We are doing it right, in which case there is no "unit cost" for APD key personnel in this table, there is only a total cost. (We could compute a unit cost as `total / commitment`, but I chose not to do that because in the table, it looks like a user input but it's calculated. It's also not a useful number to calculate since it is only for display in this one table.)

2. We are doing it wrong and the cost entered in the key personnel section is that person's total cost to the _state_, not to the program, and we should scale their cost based on their time commitment to the program.

In either case, we probably ought to update the help text to be clear which we're asking for. (And if we're going to tweak help text, can we also explain what it means for someone to have a commitment greater than 0% but also have no costs attributable to the program?)

jeromeleecms commented:
Makes sense - I'm inclined to leave this as is and spin up a separate issue to address this potential change since it does seem to be going out of scope of scope for this particular Issue/PR.

Does that work?

mgwalker commented:
@jeromeleecms 
1 and 2: Fixed up the budget summary narrative at the bottom.
3. Nothing to change I think?
4: Tightened up the columns some more so that they take up less extra space, which resulted in better alignment with the header!
5: Punting to a separate issue
6: Removed the blank cell in all those tables since that's not a proper way of getting padding, and instead added padding. Now they all have the same amount of padding, so the first columns all line up!
